### PR TITLE
Assignment7+8 linq - Cole Stewart and Matthew Powers

### PR DESCRIPTION
**Pull request by colestewart12**
@colestewart12 @Matthew-Powers

### CODE REVIEW

**Comment by EthanAGi** [comment]
Reviewing!

File: Assignment.Tests/NodeTests.cs [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+        cars.Append("Toyota");
+        cars.Append("Dodge");
+        cars.Append("Honda");
+
+        List<string> expectedResult = ["BMW", "Volvo", "Ford"];
+
+        var result = cars.ChildItems(3).ToList();
+
+        Assert.IsTrue(expectedResult.SequenceEqual<string>(result.Select(node => node.Value)));
+    }
```
Comment (EthanAGi) [inline Assignment.Tests/NodeTests.cs:42]: This test is really interesting, the Assert is really cool since it's so complex!

File: Assignment.Tests/SampleDataTests.cs [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+[TestClass]
+public class SampleDataTests
+{
+    [TestMethod]
+    public void CsvRows_ShouldSkipHeaderAndReturnCorrectLines()
+    {
+        // Arrange
+        var sampleData = new SampleData();
+
```
Comment (EthanAGi) [inline Assignment.Tests/SampleDataTests.cs:13]: While this is probably fine, you could make an explicit call here instead of using var.

File: Assignment.Tests/SampleDataTests.cs [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        // Act
+        var result = sampleData.GetAggregateListOfStatesGivenPeopleCollection(people);
+        Console.WriteLine(result[0] + "" + result[1]);
+        Console.WriteLine(csvResult[0]);
+
+        // Assert
+        Assert.AreEqual(csvResult[0], result[0] + "" + result[1]);
+    }
+}
```
Comment (EthanAGi) [inline Assignment.Tests/SampleDataTests.cs:92]: I really like your tests, my only criticism being mentioned earlier, where some explicit calls can be made instead.

File: Assignment/SampleData.cs [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+            while (!reader.EndOfStream)
+            {
+                string? line = reader.ReadLine();
+                if (line != null)
+                {
+                    yield return line;
+                }
+            }
+        }
```
Comment (EthanAGi) [inline Assignment/SampleData.cs:33]: This isn't bad, but you could use LINQ to cut this down significantly, using something like "File.ReadLine(CsvPath).Skip(1)" could do a lot of the work here!

File: Assignment/SampleData.cs [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+    // 2.
+    public IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows()
+    {
+        return CsvRows
+            .Select(row => row.Split(',')[6])
+            .Distinct()
+            .OrderBy(state => state);
+    }
+
```
Comment (EthanAGi) [inline Assignment/SampleData.cs:44]: In our code, we had to use the index 4 to find the state, I'm not sure how yours works this way but good job!

File: Assignment/SampleData.cs [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    // 3.
+    public string GetAggregateSortedListOfStatesUsingCsvRows()
+    {
+        return string.Join(", ", GetUniqueSortedListOfStatesGivenCsvRows());
+    }
+
```
Comment (EthanAGi) [inline Assignment/SampleData.cs:50]: Since yours works, it may not be required, but the README requires a ".ToArray()".

**Review by EthanAGi — COMMENTED** [review]
Hey! Pretty good! I noted some things I saw, but otherwise you guys did a great job!

Implement the ISampleData.CsvRows property, loading the data from the People.csv file and returning each line as a single string. ✔
Change the "Copy to" property on People.csv to "Copy if newer" so that the file is deployed along with your test project. ✔
Using LINQ, skip the first row in the People.csv. ❌<- I believe the comment I left on this area explains how to do this using “.Skip(1)”
Be sure to appropriately handle resource (IDisposable) items correctly if applicable (and it may not be depending on how you implement it). ✔
Implement IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows() to return a sorted, unique list of states. ✔
Use ISampleData.CsvRows for your data source. ✔
Don't forget the list should be unique. ✔
Sort the list alphabetically. ✔
Include a test that leverages a hardcoded list of addresses. ❌<- Couldn’t find a hard coded list of Addresses, unless People.csv counts.
Include a test that uses LINQ to verify the data is sorted correctly (do not use a hardcoded list). ✔
Implement ISampleData.GetAggregateSortedListOfStatesUsingCsvRows() to return a string that contains a unique, comma separated list of states. ✔
Use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows() for your data source. ✔
Consider "selecting" only the states and calling ToArray() to retrieve an array of all the state names. ❌< - No “.ToArray()” included.
Given the array, consider using string.Join to combine the list into a single string. ✔
Implement the ISampleData.People property to return all the items in People.csv as Person objects ✔
Use ISampleData.CsvRows as the source of the data. ✔
Sort the list by State, City, and Zip. (Sort the addresses first then select). ✔
Be sure that Person.Address is also populated. ✔
Adding null validation to all the Person and Address properties is optional.
Consider using ISampleData.CsvRows in your test to verify your results. ❌ <- Couldn't find any “CsvRows”
Implement ISampleDate.FilterByEmailAddress(Predicate<string> filter) to return a list of names where the email address matches the filter. ✔
Use ISampleData.People for your data source. ✔
Implement ISampleData.GetAggregateListOfStatesGivenPeopleCollection(IEnumerable<IPerson> people) to return a string that contains a unique, comma-separated list of states. ✔
Use the people parameter from ISampleData.People property for your data source. ✔
At a minimum, use the System.Linq.Enumerable.Aggregate` LINQ method to create your result. ✔
Don't forget the list should be unique. ✔
It is recommended that, at a minimum, you use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows to validate your result.
Given the implementation of Node in Assignment5
Implement IEnumerable<T> to return all the items in the "circle" of items. ✔
Add an IEnumberable<T> ChildItems(int maximum) method to Node that returns the remaining items with a maximum number of items returned less than maximum.✔
Fundamentals
Place all shared project properties into a Directory.Build.Props file.✔
Place all shared project items into a Directory.Build.targets file.✔
Ensure nullable reference types is enabled ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
and enabled .NET analyzers for both projects ✔
For this assignment, consider using Assert.AreEqual<T>() (the generic version) ✔
All of the above should be unit tested ✔
Choose simplicity over complexity ✔

**Comment by Anna-Mclenko** [comment]
Ill review this too!

File: Assignment/SampleData.cs (hunk 4 above)
Comment (Anna-Mclenko) [reply Assignment/SampleData.cs:33]: Yes I agree! we also used Linq in our assugnment.

File: Assignment/SampleData.cs (hunk 5 above)
Comment (Anna-Mclenko) [reply Assignment/SampleData.cs:44]: I like what you did here! simple, concise, gets the job done! :)

File: Assignment.Tests/SampleDataTests.cs [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+public class SampleDataTests
+{
+    [TestMethod]
+    public void CsvRows_ShouldSkipHeaderAndReturnCorrectLines()
+    {
+        // Arrange
+        var sampleData = new SampleData();
+
+        // Act
+        var csvRows = sampleData.CsvRows.Count();
```
Comment (Anna-Mclenko) [inline Assignment.Tests/SampleDataTests.cs:15]: Yeah, I think actual types will be better than 'var' like Ethan said

File: Assignment/SampleData.cs (hunk 6 above)
Comment (Anna-Mclenko) [reply Assignment/SampleData.cs:50]: me and my teammate also used the .toArray() in our code. Though the instructions do say "consider" so i guess we didnt need to  use it... i guess ill give you points for that :)

File: Assignment/SampleData.cs [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+                {
+                    var columns = row.Split(',');
+                    var address = new Address(columns[4], columns[5], columns[6], columns[7]);
+                    return new Person(columns[1], columns[2], address, columns[3]);
+                })
+                .OrderBy(person => person.Address.State)
+                .ThenBy(person => person.Address.City)
+                .ThenBy(person => person.Address.Zip);
+        }
+    }
```
Comment (Anna-Mclenko) [inline Assignment/SampleData.cs:67]: OOOOOH! i like how you did it here! we did 2 selects the first one split everything then we sorted all the rows then the last select created the person object. But this is a nice way to do it!

File: Assignment.Tests/SampleDataTests.cs [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+
+        // Assert
+        Assert.AreEqual(50, csvRows);
```
Comment (Anna-Mclenko) [inline Assignment.Tests/SampleDataTests.cs:18]: why arent you using the generic version? 
`Assert.AreEqual<int>(50, csvRows)`
i think will be much better

**Review by Anna-Mclenko — COMMENTED** [review]
Instructions

Throughout, consider using the System.Linq.Enumerable methods Zip, Count, Sort and Contains methods for testing collections.. (Preferably avoid using Microsoft.VisualStudio.TestTools.UnitTesting.CollectionAssert although that might be easier, to get a firmer grasp on additional LINQ API.)

Implement the ISampleData.CsvRows property, loading the data from the People.csv file and returning each line as a single string. ✔

Change the "Copy to" property on People.csv to "Copy if newer" so that the file is deployed along with your test project. ✔

Using LINQ, skip the first row in the People.csv. ❌

Be sure to appropriately handle resource (IDisposable) items correctly if applicable (and it may not be depending on how you implement it). ✔

Implement IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows() to return a sorted, unique list of states. ✔

Use ISampleData.CsvRows for your data source. ✔

Don't forget the list should be unique. ✔

Sort the list alphabetically. ✔

Include a test that leverages a hardcoded list of addresses. ❌

Include a test that uses LINQ to verify the data is sorted correctly (do not use a hardcoded list). ✔

Implement ISampleData.GetAggregateSortedListOfStatesUsingCsvRows() to return a string that contains a unique, comma separated list of states. ✔

Use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows() for your data source. ✔

Consider "selecting" only the states and calling ToArray() to retrieve an array of all the state names. ✔

Given the array, consider using string.Join to combine the list into a single string. ✔

Implement the ISampleData.People property to return all the items in People.csv as Person objects ✔

Use ISampleData.CsvRows as the source of the data. ✔

Sort the list by State, City, and Zip. (Sort the addresses first then select). ✔

Be sure that Person.Address is also populated. ✔

Adding null validation to all the Person and Address properties is optional.
Consider using ISampleData.CsvRows in your test to verify your results. ✔ - do we have to use CsvRows? They did a pretty good job testing without it I think

Implement ISampleDate.FilterByEmailAddress(Predicate<string> filter) to return a list of names where the email address matches the filter. ✔

Use ISampleData.People for your data source. ✔

Implement ISampleData.GetAggregateListOfStatesGivenPeopleCollection(IEnumerable<IPerson> people) to return a string that contains a unique, comma-separated list of states. ✔

Use the people parameter from ISampleData.People property for your data source. ✔

At a minimum, use the System.Linq.Enumerable.Aggregate` LINQ method to create your result. ✔

Don't forget the list should be unique. ✔

It is recommended that, at a minimum, you use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows to validate your result.
Given the implementation of Node in Assignment5

Implement IEnumerable<T> to return all the items in the "circle" of items. ✔

Add an IEnumberable<T> ChildItems(int maximum) method to Node that returns the remaining items with a maximum number of items returned less than maximum.✔

Extra Credit
Implement the homework using async/await and multi-threading by defining a new SampleDataAsync class that implements IAsyncSampleData). Refactor your SampleData and SampleDataAsync classes with minimal duplication. Be sure to refactor your tests to re-use a significant amount of the test code for both implementations. ❌

Fundamentals

Place all shared project properties into a Directory.Build.Props file.✔
Place all shared project items into a Directory.Build.targets file.✔

Ensure nullable reference types is enabled ✔

Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔

Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔

and enabled .NET analyzers for both projects ✔

For this assignment, consider using Assert.AreEqual<T>() (the generic version) ❌

All of the above should be unit tested ✔

Choose simplicity over complexity ✔

File: Assignment.Tests/SampleDataTests.cs (hunk 2 above)
Comment (Matthew-Powers) [reply Assignment.Tests/SampleDataTests.cs:13]: Will fix

File: Assignment/SampleData.cs (hunk 4 above)
Comment (Matthew-Powers) [reply Assignment/SampleData.cs:33]: Will fix

File: Assignment.Tests/SampleDataTests.cs (hunk 9 above)
Comment (Matthew-Powers) [reply Assignment.Tests/SampleDataTests.cs:18]: IntelliSense says that it can be simplified to what it is now, rather than having the <int>

File: Assignment.Tests/NodeTests.cs (hunk 1 above)
Comment (Matthew-Powers) [reply Assignment.Tests/NodeTests.cs:42]: Thanks

File: Assignment.Tests/SampleDataTests.cs (hunk 3 above)
Comment (Matthew-Powers) [reply Assignment.Tests/SampleDataTests.cs:105]: Fixed the issues

File: Assignment/SampleData.cs (hunk 5 above)
Comment (Matthew-Powers) [reply Assignment/SampleData.cs:27]: Thanks

File: Assignment/SampleData.cs (hunk 6 above)
Comment (Matthew-Powers) [reply Assignment/SampleData.cs:33]: Given that it says consider rather than stated as a requirement, will be keeping code the same

File: Assignment.Tests/SampleDataTests.cs (hunk 7 above)
Comment (Matthew-Powers) [reply Assignment.Tests/SampleDataTests.cs]: Fixed

File: Assignment/SampleData.cs (hunk 8 above)
Comment (Matthew-Powers) [reply Assignment/SampleData.cs:50]: Thanks

File: Assignment/Node.cs [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+
+        while(currentNode.Next != this)
+        {
+            yield return currentNode;
+            currentNode = currentNode.Next;
+        }
+        if(currentNode.Next == this)
+        {
+            yield return currentNode;
+        }
```
Comment (BenjaminMichaelis) [inline Assignment/Node.cs:80]: What's the reason for `this` being returned last vs first?

File: Assignment.Tests/NodeTests.cs [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+        Node<string> cars = new("BMW");
+        cars.Append("Volvo");
+        cars.Append("Ford");
+        cars.Append("Toyota");
+        cars.Append("Dodge");
+        cars.Append("Honda");
+
+        List<string> expectedResult = ["BMW", "Volvo", "Ford"];
+
+        List<Node<string>> result = cars.ChildItems(3).ToList();
```
Comment (BenjaminMichaelis) [inline Assignment.Tests/NodeTests.cs:39]: There could be a good argument made for whether maximum should be inclusive or exclusive. Also I am surprised that "ChildItems" would return itself and not just it's children.

File: Assignment.Tests/NodeTests.cs [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+        while (result.MoveNext())
+        { 
+            count++; 
+        }
+
+        Assert.AreEqual(3, count);
+    }
+
+    [TestMethod]
+    public void NodeTests_ReturnThreeItems_Success()
```
Comment (BenjaminMichaelis) [inline Assignment.Tests/NodeTests.cs:28]: What about edge case tests? what if I pass in `-4`?

File: Assignment.Tests/SampleDataTests.cs [hunk 13]
```diff
@@ -0,0 +1,105 @@
+﻿using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+namespace Assignment.Tests;
+
+[TestClass]
+public class SampleDataTests
```
Comment (BenjaminMichaelis) [inline Assignment.Tests/SampleDataTests.cs:6]: These seem like a lot of "pinning" tests, basically making sure the existing functionality is as you would expect. I would encourage you to make sure you cover edge cases, and think of ways that failures may occur.
